Useful Starting Page backend changes

adam-james-v opened the pull request:
Epic: #21322 

#### BE changes

- [x] Add `is_installer` field to `/api/user/current`. It's basically the user with `id = 1`. Make sure to update `add-has-invited-second-user` check.
- [x] Add `has_question_and_dashboard` to `/api/user/current`. Should be `true` if the current user has access to at least one question *and* dashboard.
- [x] Add `/api/activity/popular_views` using the logic outlined in [the product doc](https://www.notion.so/metabase/Give-users-a-chill-and-useful-starting-page-41fce022e6c448bd8036d4facec7eb51#1e39bc3549a547b4802654fda2fb1e09). Could return the same data as from `/api/activity/recent_views`.
- [ ] (Optional) Add `first_login` to `/api/user/current`. If not present in the db, this field can be substituted with `date_joined` for existing users. The field mirrors `last_login` field.

adam-james-v committed:
Add `:is_installer` key to users returned from api/user/current

adam-james-v committed:
Add `:is_installer` key to users returned from api/user/current

adam-james-v committed:
Add `:has_question_and_dashboard` Key to current users via hydration

This is also part of the starting page project, it's a key that helps the frontend serve more specific data to the
user.

adam-james-v committed:
Moved `permissions-set` function up so I can use it in hydration

adam-james-v committed:
Moved `permissions-set` function up so I can use it in hydration

kulyk committed:
Browsing | Refactor navigation sidebar (#21298)

* Create sidebar and add collections tree

* Fix dragged collection item's z-index

* Add links (data, archive, users' collections)

* Add ProfileLink

* Add bookmarks

* Use the new sidebar at Navbar

* Fix official collection icon opacity

* Add sidebar width to navigation constants

* Cut long item names

* Fix sidebar's horizontal scroll

* Fix remove bookmark icon position

* Fix bookmark URLs

* Fix how selected item is determined

* Jump to a collection opened from a bookmark

* Fix selected element highlighting edge cases

* Don't check for "users" collection

* Fix collapsing tree nodes

* Fix remove bookmark icon show on hover

* Fix removing a bookmark

* Fix archived collections are disabled in the tree

* Expand collection's parents after move

* Fix UI element lookups in tests

* Accept `icon` object as a SidebarLink prop

* Use "tree" a11y role for collections tree

* Always show the archive link

* Fix "other users' collections" link icon

* Fix e2e test

* Fix e2e tests

* Remove not used prop

* Make bookmarks collapsible

* Fix post-rebase issues

* Fix can't click expanded tree node to collapse

* Fix e2e test

* Remove not used code

* Use bookmark entities icon helper

* Extract renderBookmark function

* Allow custom `icon` element prop on SidebarLink

* Fix bookmarked official collection icon colors

* Fix `personal_owner_id` type

* Move Collections entity file to collections dir

* Reorganize collection entity file

Commit:
Downgrade buildJet instance of Percy (#21391)

* Moving to a 4vCPU

* New tests

* Test Finish

daltojohnso committed:
@sugengsuprayogi's "Fix download question data on embedded dashboard" (#21231)

* Fix download question data on embedded dashboard

* run prettier

Co-authored-by: Sugeng Suprayogi <sugengsuprayogi@users.noreply.github.com>

nemanjaglumac committed:
Set `modifyObstructiveCode` to false in Cypress config (#21376)

kulyk committed:
Browsing | Improve mobile sidebar (#21347)

nemanjaglumac committed:
[E2E] Reinforce repro for #17514 (#21394)

* Reinforce the repro for 17514

* Fix typos

camsaul committed:
Skip `filter-by-current-quarter-test` for :oracle because of #10072 (#21396) [ci skip]

daltojohnso committed:
Set visualization mode via QB state when we are zoomed in on a specific row (#21371)

* Set visualization mode via QB state

* Unskip repro

kulyk committed:
Browsing | Show sidebar on collection pages when embedding (#21352)

kulyk committed:
Browsing | Automatically close navbar in certain scenarios (#21358)

kulyk committed:
[E2E] Fix flaky collections sidebar tests (#21397)

gusaiani committed:
Update bookmark list when archiving Questions in Collection page (#21390)

npfitz committed:
Updating page title as dashboard loads (#21259)

Updating Page Title and Favicon as Dashboards load

kulyk committed:
Remove collection sidebar code (#21400)

kulyk committed:
Remove `getPlainNativeQuery` (#21401)

kulyk committed:
Clean up EditBar component (#21395)

losrebellos committed:
15876 fix time random casting with without timezone (#21340)

* Misc

* parseTime method updated to remove timezone

* parseTime tests added

* formatTime tests added

* Unskip 15876 cypress test

kulyk committed:
Don't autoexpand sidebar on mobile (#21415)

nemanjaglumac committed:
Fix flaking Percy `table` snapshot in CCI [ci skip] (#21416)

nemanjaglumac committed:
Fix flaking collection tests [ci skip] (#21417)

Commit:
Ignore running E2E workflow when not needed (#21384)

* Ignore running E2E workflow when not needed

* Adding - ".**" as suggested

* Playing safe for now

noahmoss committed:
Data model permission writes & reads (#21168)

noahmoss committed:
DB details permissions reads & writes (#21327)

gusaiani committed:
Cover more entities (#21406)

qnkhuat committed:
Grant Subscription permission by default (#21179)

* add API to fetch general permisisons graph

* add API to update general permissionns

* change author of migration

* update documents

* misc fixes to applease the CIs

* Add tests for general permission APIs and models

* linting and fix a failed test case

* fix some failed tests

* update docs and change /subscription/ to /general/subscription/ for consistency

* Hook and migration to make sure subscription are created for new groups by default

* add schema migrations tests

* set for the win

* address noah's comments

* Parse number as is in http-client for tests

* address Cam's comments

* revert the last commit about parsing API response in tests

* change fk name

* delete a comment

* Changes:
- Rename `changes` column to `after` to keep things consistent
- If a group  has no General Permisions, it'll not be included in the graph
- Update tests and some docs

* fix failing tests in ee

* add some tests and make docstring completes

* polishing comments

* namespaces

* fix namespaces

* Add general permisison flags to `GET /api/user/current` (#21250)

* return general permisison flags for /api/current

* namespaces

* move permission flags to under

* Enforce Subscription permissions (#21285)

* return general permisison flags for /api/current

* namespaces

* enforce general permissions for subscription and tests

* update check-has-general-permisison to add option to require superuser

* adding arity

* add tests for permissions helper function

* Move advanced permissions check funcs to ee namespace

* unpushed changes

* namespaces

* ignore exception when load namespaces

* change helper fn name

* Enforce Monitoring Permissions (#21321)

* return general permisison flags for /api/current

* namespaces

* enforce general permissions for subscription and tests

* update check-has-general-permisison to add option to require superuser

* adding arity

* enforce permissions to call /api/dataset for internal queries

* enforce monitoring permissions for api/task and api/util

* add tests for OSS

* add tests for db-connection-info endpoint

* change test schema

* update name func and fix ns

* whydon't CI run ?

* Enforce Setting Permissions (#21386)

* return general permisison flags for /api/current

* namespaces

* enforce general permissions for subscription and tests

* update check-has-general-permisison to add option to require superuser

* adding arity

* enforce permissions to call /api/dataset for internal queries

* enforce monitoring permissions for api/task and api/util

* add tests for OSS

* add tests for db-connection-info endpoint

* change test schema

* update name func and fix ns

* whydon't CI run ?

* Enforce Setting permissions

* fix failing test

* make sure we could run slack test twice

* make the mock consistent

* address Noah's comments

* shorter permissions check

Commit:
Fix name [skip ci] (#21424)

nemanjaglumac committed:
[E2E] Fix flake in repro 17514 for good (#21423)

ranquild committed:
The skeleton of the new homepage (#21355)

adam-james-v committed:
Adjust recents query and first pass at popular items query

Recents:

Before, the recent query would look at the ViewLog, which records a view on cards even when it was only 'viewed' via a
dashboard.

For dashboards and tables, we don't need to make a change.

For cards, we now query the QueryExecution table and use started_at as a proxy for the last viewed timestamp, and can
then only grab cards with an execution context of 'question'.

Popular:

We now want to have a notion of 'popular' items as well, which can look different than 'recents'. This is the first
attempt at a reasonable query with scoring system.

The score takes into account:
 - recency of view
 - count of views total
 - if the card/dashboard is 'verified' (enterprise)
 - if the cad/dashboard is inside an 'official' Collection (enterprise)

The popular score currently uses only the *current-user-id* to fetch items, so popularity is (in this commit, at
least) a per-user concept.

adam-james-v committed:
Adjust recents query and first pass at popular items query

Recents:

Before, the recent query would look at the ViewLog, which records a view on cards even when it was only 'viewed' via a
dashboard.

For dashboards and tables, we don't need to make a change.

For cards, we now query the QueryExecution table and use started_at as a proxy for the last viewed timestamp, and can
then only grab cards with an execution context of 'question'.

Popular:

We now want to have a notion of 'popular' items as well, which can look different than 'recents'. This is the first
attempt at a reasonable query with scoring system.

The score takes into account:
 - recency of view
 - count of views total
 - if the card/dashboard is 'verified' (enterprise)
 - if the cad/dashboard is inside an 'official' Collection (enterprise)

The popular score currently uses only the *current-user-id* to fetch items, so popularity is (in this commit, at
least) a per-user concept.

adam-james-v committed:
Fixed mistake and renamed endpoint

adam-james-v committed:
Fixed mistake and renamed endpoint

adam-james-v committed:
Typo fix

adam-james-v committed:
Typo fix

adam-james-v committed:
Clean up QueryExecution in tests

adam-james-v committed:
Clean up QueryExecution in tests

dpsutton commented on src/metabase/api/activity.clj:
```diff
@@ -86,19 +90,33 @@
+                     :display_name :initial_sync_status
+                     :visibility_type])
+                    (cond-> {:where [:in (db/qualify (symbol (str/capitalize model)) :id) ids]}
+                      (not= model "table")
```
`(not= model "table')` is inside of the clause when model is equal to table.

dpsutton replied:
ah. the indentation here is off so it isn't clear that the `cond->` is the argument to the `db/select`

dpsutton replied:
you don't need the `(db/qualify Dashboard :id)` or `(db/qualify Card :id)` (and note you don't have it on the `Table`'s id.

adam-james-v committed:
Add `:has_question_and_dashboard` Key to current users via hydration

This is also part of the starting page project, it's a key that helps the frontend serve more specific data to the
user.

dpsutton commented on src/metabase/api/activity.clj:
```diff
@@ -107,31 +125,122 @@
-                                     [:in :model #{"card" "dashboard" "table"}]]
-                          :order-by [[:max_ts :desc]]
-                          :limit    8})
+  (let [views (views-and-runs 8 8)
```
normally want to create some defs for these two `8`s so its clear what they are and where to tweak them if necessary.

adam-james-v replied:
I was considering using keyword args to make it clear, but keep the values at the call sites?

I could also make top level defs, like you say. If I do that, should I make `views-and-runs` a 0-arity fn?

dpsutton replied:
yeah i would just `(def last-view-count 8)` or something similar. Verbose is ok. Just make it easy for those constants to stand out if anyone needs to change in the future.

adam-james-v committed:
Merge branch 'start-page-be' of https://github.com/metabase/metabase into start-page-be

dpsutton commented on src/metabase/models/user.clj:
```diff
@@ -165,6 +184,32 @@
+                         (collection/permissions-set->visible-collection-ids (permissions-set user)))]})]
+    (when (seq users)
+      (for [user users]
+        (assoc user :has_question_and_dashboard (and (> (db/count 'Card (perms-query user)) 0)
```
Two thoughts:
- `perms-query` makes two db calls each time and you invoke it twice instead of just once and reusing it
- `perms-query` makes two calls and we do it for each user. This could get really expensive depending on the number of users. Is it possible to batch this up and get some map of `user-id->permissions`?

adam-james-v replied:
Oh, that's a good point. I'll rework this, thanks

dpsutton commented on test/metabase/api/activity_test.clj:
```diff
@@ -96,6 +98,21 @@
+(defn- create-runs!
+  "Insert views [user-id model model-id]. Reviews are entered a second apart with last review as most recent."
+  [runs]
+  (let [runs (map (fn [[user _model model-id] hours-ago]
```
`hours-ago` seems like a bad name for `(t/seconds (- hours-ago))`

dpsutton reviewed (changes requested):
looks like a good avenue. But the N+1 style db queries for multiple users is too much. I think I saw in slack that you are limiting this to just the current user for the moment. If that's the case you might be interested in 
```clojure
metabase.api.common/*current-user-permissions-set*
  Delay to the set of permissions granted to the current user. See documentation in [[metabase.models.permissions]] for
  more information about the Metabase permissions system.
```
Until we get a better way to get permissions for a set of users.

adam-james-v committed:
Indent, naming, and some simple defs

adam-james-v committed:
try to reduce db calls for :has_question_and_dashboard

I've moved the fn out of the hydration system to guarantee that it only runs for a single user, and is only used in
`GET /api/user/current` (and not accidentally used in other spots via hydration mechanism).

Still don't know how to check Card and Dashboard tables more efficiently
